chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/TestToolReviewer.java

```diff
@@ -19,37 +23,88 @@
 public class TestToolReviewer extends PermissionBasedReviewer {
     private TestToolDAO testToolDao;
 
+    private List<TestToolCriteriaMap> testToolCriteriaMap;
+
     @Autowired
-    public TestToolReviewer(TestToolDAO testToolDAO, ErrorMessageUtil msgUtil, ResourcePermissions resourcePermissions) {
+    public TestToolReviewer(TestToolDAO testToolDAO, ErrorMessageUtil msgUtil, ResourcePermissions resourcePermissions) throws EntityRetrievalException {
         super(msgUtil, resourcePermissions);
         this.testToolDao = testToolDAO;
+
+        testToolCriteriaMap = testToolDao.getAllTestToolCriteriaMap();
     }
 
     @Override
     public void review(CertifiedProductSearchDetails listing) {
-        for (CertificationResult cert : listing.getCertificationResults()) {
-            if (cert.isSuccess() != null && cert.isSuccess().equals(Boolean.TRUE)) {
-                if (cert.getTestToolsUsed() != null && cert.getTestToolsUsed().size() > 0) {
-                    Iterator<CertificationResultTestTool> testToolIter = cert.getTestToolsUsed().iterator();
-                    while (testToolIter.hasNext()) {
-                        CertificationResultTestTool testTool = testToolIter.next();
-                        if (StringUtils.isEmpty(testTool.getTestToolName())) {
-                            addCriterionErrorOrWarningByPermission(listing, cert, "listing.criteria.missingTestToolName",
-                                    Util.formatCriteriaNumber(cert.getCriterion()));
-                        } else {
-                            TestToolDTO tt = testToolDao.getByName(testTool.getTestToolName());
-                            if (tt != null && tt.isRetired()) {
-                                listing.getWarningMessages()
-                                        .add(msgUtil.getMessage("listing.criteria.retiredTestToolNotAllowed",
-                                                testTool.getTestToolName(), Util.formatCriteriaNumber(cert.getCriterion())));
-                            } else if (tt == null) {
-                                addCriterionErrorOrWarningByPermission(listing, cert, "listing.criteria.testToolNotFound",
-                                        Util.formatCriteriaNumber(cert.getCriterion()), testTool.getTestToolName());
-                            }
-                        }
-                    }
-                }
+        listing.getCertificationResults().stream()
+                .filter(cr -> isCertificationResultAttestedTo(cr)
+                        && doesCertificationResultHaveTestTools(cr))
+                .forEach(cr -> validateTestTools(listing, cr));
+    }
+
+    private Boolean isCertificationResultAttestedTo(CertificationResult cert) {
+        return cert.isSuccess() != null && cert.isSuccess().equals(Boolean.TRUE);
+    }
```

Comment:
Is this a method we could lift to a parent class? It's not specific to Test Tools, and seems like it would be useful on all of our criterion reviewers